middleware/uncompress.go

```diff
@@ -0,0 +1,44 @@
+package middleware
+
+import (
+	"compress/gzip"
+	"io"
+	"net/http"
+)
+
+// UncompressMiddleware uncompresses gzipped HTTP requests carrying a 'Content-Encoding: gzip' header.
+var UncompressMiddleware = func(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Content-Encoding") == "gzip" {
+			r.Body = &gzipReader{body: r.Body}
+		}
+		h.ServeHTTP(w, r)
+	})
+}
+
+// gzipReader wraps a response body so it can lazily
+// call gzip.NewReader on the first call to Read
+type gzipReader struct {
+	body io.ReadCloser // underlying request body
+	zr   *gzip.Reader  // lazily-initialized gzip reader
+	zerr error         // any error from gzip.NewReader; sticky
+}
+
+func (gz *gzipReader) Read(p []byte) (n int, err error) {
+	if gz.zr == nil {
```

Comment:
Should we also check if gz.zerr ! = nil also return immediately?

Reply:
Borrowed implementation from `net/http/transport` [link](https://go.dev/src/net/http/transport.go?s=13180:13237#:~:text=func%20(gz%20*gzipReader)%20Read(p%20%5B%5Dbyte)%20(n%20int%2C%20err%20error)%20%7B)

Found it smart, since you only need to check for `zerr` once. If we check for `zerr` early then we need to do it again after `gzip.NewReader`.

The only modification I did in transport's gzipReader was to close the `gzip.Reader` if not nil during `Close` method.